pkg/hooks/upgrade_hooks.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hooks
+
+import (
+	"fmt"
+	"reflect"
+	"runtime"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/hashicorp/go-version"
+
+	"istio.io/operator/pkg/apis/istio/v1alpha2"
+	"istio.io/operator/pkg/util"
+	"istio.io/pkg/log"
+)
+
+// ExecClient is an interface for remote execution
+// TODO: replace with correct type.
+type ExecClient interface {
+	PodsForSelector(namespace, labelSelector string) (*v1.PodList, error)
+}
+
+// hook is a callout function that may be called during an upgrade to check state or modify the cluster.
+// hooks should only be used for version-specific actions.
+type hook func(kubeClient ExecClient, sourceValues, targetValues *v1alpha2.IstioControlPlaneSpec) util.Errors
+type hooks []hook
+
+// hookVersionMapping is a mapping between a hashicorp/go-version formatted constraints for the source and target
+// versions and the list of hooks that should be run if the constraints match.
+type hookVersionMapping struct {
+	sourceVersionConstraint string
+	targetVersionConstraint string
+	hooks                   []hook
+}
+
+// hookCommonParams is a set of common params passed to all hooks.
+type hookCommonParams struct {
+	sourceVer    string
+	targetVer    string
+	sourceValues *v1alpha2.IstioControlPlaneSpec
+	targetValues *v1alpha2.IstioControlPlaneSpec
+}
+
+var (
+	// preUpgradeHooks is a list of hook version constraint pairs mapping to a slide of corresponding hooks to run
+	// before upgrade.
+	preUpgradeHooks = []hookVersionMapping{
+		{
+			sourceVersionConstraint: "1.3",
+			targetVersionConstraint: "",
+			hooks:                   []hook{checkInitCrdJobs},
+		},
```

Comment:
I am thinking abstract this into an interface, so the hook can be a self-contained type and easy to implement by others without understanding the code here.

Reply:
yes, it could be done like this. my only concern would be if we allow hooks to be registered from anywhere, the code may become quite scattered. keeping all this private (except runPreUpgradeHooks and runPostUpgradeHooks, which should be public) has the advantage that all these hacks are contained to the hooks package.